# Problem Statement

By replacing the 1st digit of the 2-digit number *3, it turns out that six of the nine possible values: 13, 23, 43, 53, 73, and 83, are all prime.

By replacing the 3rd and 4th digits of 56\*\*3 with the same digit, this 5-digit number is the first example having seven primes among the ten generated numbers, yielding the family: 56003, 56113, 56333, 56443, 56663, 56773, and 56993. Consequently 56003, being the first member of this family, is the smallest prime with this property.

Find the smallest prime which, by replacing part of the number (not necessarily adjacent digits) with the same digit, is part of an eight prime value family.

## Solution 1

Gonna brute force this for sure, but I'm going to limit the search space first. First, note that the last digit of any $n$ digit number will never be up for replacement, because half of the numbers will be divisble by 2. I.e. if we have some number of the form:

$N = abcd..z*$

Then when $* \in \{2, 4, 6, 8, 0\}$ $N$ will be divisble by 2.

It follows naturally, that the last digit of any $N$ must be in the set $D = \{1, 3, 7, 9\}$.

### One Digit Replacement

Consider numbers of the form $nd$ where $d \in D$. I will show there cannot be a family of primes with 8 members in the two digit numbers using the 3's divisibility rule. The sum $n_{1} + d$ will be divisible by 3 at least 3 times no matter what the value of $d$ is. No need to check anything here. Similarly, we can never only replace a single number, otherwise the value of the sum of digits will be divisble by 3 at least 3 times.

### Two Digit Replacement
From the previous problem we can ignore numbers where we only replace a single digit, so the only valid form must be if we replace both of the first two digits. So, consider numbers of the form $nnd$ where $d \in D$. Consider the set of sums generated by adding the two digits $n$:

$S = \{0, 2, 4, 6, 8, 10, 12, 14, 16, 18\}

If we add our digit $d$ to this set, no matter the value, at least 3 elements will be divisible by $3$. Since we're only interested in divisibility by 3, we'll take the set $S$ mod 3:

$$S' = \{0, 2, 1, 0, 2, 1, 0, 2, 1, 0\}$$

We can see that no matter the number we can't replace only two digits and get a family of 8 primes, since at least 3 numbers will be divisible.

### k Digit Replacement

Based on the previous replacements, we can extend it out by adding any multiple of 3. That is to say that we cannot get 8 primes unless we replace $k$ digits where $k | 3$. This means that the residual set mod 3 will always be 0. And Thus, $d$ must actuall be [1, 9]



In [4]:
import os
import sys

module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)

from utils import prime_utils

is_prime, primes = prime_utils.load_primes()
ds = [1, 3, 7, 9]

In [17]:
# Check 4 digit numbers.
for d in ds:
    num_primes = 0
    for i in range(10):
        if is_prime[int(f"{i}{i}{i}{d}")]:
            num_primes += 1
    print(f"***{d}: {num_primes}")

***1: 4
***3: 1
***7: 5
***9: 0


In [20]:
# Check 5 digit numbers.
ds = [1, 3, 7, 9]
for d in ds:
    valid_ns = []
    for n in range(10):
        if (d + n) % 3 != 0:
            valid_ns.append(n)
    for n_index in [0, 1, 2, 3]:
        format_list = ["{0}"] * 4 + [str(d)]
        for n in valid_ns:
            format_list[n_index] = str(n)
            num_primes = 0
            for i in range(10):
                if is_prime[int("".join(format_list).format(i))]:
                    num_primes += 1
            print("".join(format_list), num_primes)
    
    


0{0}{0}{0}1 4
1{0}{0}{0}1 4
3{0}{0}{0}1 1
4{0}{0}{0}1 3
6{0}{0}{0}1 4
7{0}{0}{0}1 4
9{0}{0}{0}1 6
{0}0{0}{0}1 3
{0}1{0}{0}1 2
{0}3{0}{0}1 5
{0}4{0}{0}1 3
{0}6{0}{0}1 3
{0}7{0}{0}1 3
{0}9{0}{0}1 4
{0}{0}0{0}1 2
{0}{0}1{0}1 4
{0}{0}3{0}1 5
{0}{0}4{0}1 3
{0}{0}6{0}1 4
{0}{0}7{0}1 3
{0}{0}9{0}1 3
{0}{0}{0}01 5
{0}{0}{0}11 5
{0}{0}{0}31 4
{0}{0}{0}41 2
{0}{0}{0}61 5
{0}{0}{0}71 4
{0}{0}{0}91 4
1{0}{0}{0}3 2
2{0}{0}{0}3 3
4{0}{0}{0}3 6
5{0}{0}{0}3 4
7{0}{0}{0}3 4
8{0}{0}{0}3 3
{0}1{0}{0}3 4
{0}2{0}{0}3 6
{0}4{0}{0}3 4
{0}5{0}{0}3 2
{0}7{0}{0}3 1
{0}8{0}{0}3 3
{0}{0}1{0}3 3
{0}{0}2{0}3 1
{0}{0}4{0}3 1
{0}{0}5{0}3 4
{0}{0}7{0}3 3
{0}{0}8{0}3 4
{0}{0}{0}13 4
{0}{0}{0}23 3
{0}{0}{0}43 5
{0}{0}{0}53 5
{0}{0}{0}73 5
{0}{0}{0}83 6
0{0}{0}{0}7 5
1{0}{0}{0}7 6
3{0}{0}{0}7 0
4{0}{0}{0}7 4
6{0}{0}{0}7 4
7{0}{0}{0}7 5
9{0}{0}{0}7 7
{0}0{0}{0}7 2
{0}1{0}{0}7 6
{0}3{0}{0}7 3
{0}4{0}{0}7 3
{0}6{0}{0}7 5
{0}7{0}{0}7 4
{0}9{0}{0}7 3
{0}{0}0{0}7 5
{0}{0}1{0}7 2
{0}{0}3{0}7 4
{0}{0}4{0}7 4
{0}{0}6{0}7 4
{0}{0}

In [5]:
import collections
import itertools

# Check 6 digit numbers.
ds = [1, 3, 7, 9]
ns = list(range(10))
qs = {
    1: collections.defaultdict(list),
    3: collections.defaultdict(list),
    7: collections.defaultdict(list),
    9: collections.defaultdict(list)
}
# Pre-generate valid numbers
for d in ds:
    for n in ns:
        for q in range(10):
            if (d + n + q) % 3 != 0:
                qs[d][n].append(q)
                
# Check our shit
for d in ds:
    for n_indices in itertools.combinations((0, 1, 2, 3, 4), 2):
        for n in ns:
            format_list = ["{0}"] * 5 + [str(d)]
            for q in qs[d][n]:
                format_list[n_indices[0]] = str(n)
                format_list[n_indices[1]] = str(q)
                num_primes = 0
                for i in range(10):
                    if is_prime[int("".join(format_list).format(i))]:
                        num_primes += 1
                if num_primes == 8:
                    for i in range(10):
                        p = int("".join(format_list).format(i))
                        if is_prime[p]:
                            print(p)
                            break

121313
857
40609
109


We get our final answer 121313 -- Assuming that the format numbers with '0's are cheating.